### ISSUE TITLE
Scan OCINum into string

### ISSUE THREAD

**Issue by omerkirk** [rana/ora#144]
Hi, I am trying to get an old web application which uses goracle into using the rana/ora driver. I got the following error. 

`sql: Scan error on column index 4: unsupported Scan, storing driver.Value type ora.OCINum into type *string]
`

I tried to change the destination into an integer, but being a legacy application I learnt that there are very good reasons why that destination should be a string. Is there any way I can get ora.OCINum to be scannable into a string destination. 

I looks like in the driver there is already a string function attached to OCINum so I couldn't figure out what is missing.

**Comment by tgulacsi** [comment]
Can you help with a minimal reproductional test case?

**Comment by tgulacsi** [comment]
I cannot reproduce
```
func TestIssue144(t *testing.T) {
    var s string
    var i int64
    var f float64
    for tN, dest := range []interface{}{
        &f,
        &i,
        &s,
    } {
        if err := testDb.QueryRow("SELECT 123456 FROM DUAL").Scan(dest); err != nil {
            t.Error("%d. %T: %v", tN, dest, err)
        }
        t.Logf("%d. %T: %v", tN, dest, reflect.ValueOf(dest).Elem().Interface())
    }
}
```

**Comment by omerkirk** [comment]
Here is the setup I am using: 
```
CREATE TABLE TEST_OCINUM_SCAN
(
  NUM                       NUMBER           NOT NULL
);

INSERT INTO TEST_OCINUM_SCAN VALUES (123456);
```

```
ora.Cfg().Env.StmtCfg.Rset.SetFloat(ora.N)
ora.Cfg().Env.StmtCfg.Rset.SetNumberFloat(ora.N)
```

Can you try it with the following test function using the above table, probably your query does not map to an ora.OCINum but to a ora.I64 or ora.I32 or ora.F64.

```
func TestIssue144(t *testing.T) {
    var s string
    var i int64
    var f float64
    for tN, dest := range []interface{}{
        &f,
        &i,
        &s,
    } {
        if err := testDb.QueryRow("SELECT NUM FROM TEST_OCINUM_SCAN").Scan(dest); err != nil {
            t.Error("%d. %T: %v", tN, dest, err)
        }
        t.Logf("%d. %T: %v", tN, dest, reflect.ValueOf(dest).Elem().Interface())
    }
}
```

**Comment by omerkirk** [comment]
There is probably very little we can do, I checked the convert.go from database/sql and types that are allowed to be converted to *string are restricted. Although I think any type that implements the Stringer interface should be convertable it doesn't seem as simple as that. The following is from the convertAssign function in convert.go:

```
case *string:
   170			sv = reflect.ValueOf(src)
   171			switch sv.Kind() {
   172			case reflect.Bool,
   173				reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
   174				reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
   175				reflect.Float32, reflect.Float64:
   176				*d = asString(src)
   177				return nil
   178			}
```

This part of the code switches over the destination type and checks the value type, which is the inner case, to see if it is convertable to *string. In my case sv.Kind() is ora.OCINum so it passes through all the cases and finally returns the following: 

`return fmt.Errorf("unsupported Scan, storing driver.Value type %T into type %T", src, dest)`

**Comment by omerkirk** [comment]
There are 4 issues at hand causing my problem here. 
1- The column that should be an INTEGER is defined as NUMBER, so it is mapped to a float64 without
 ```
ora.Cfg().Env.StmtCfg.Rset.SetFloat(ora.N)
ora.Cfg().Env.StmtCfg.Rset.SetNumberFloat(ora.N)
```
2- Unfortunately there are also columns that holds float values also so I can't simply set rset float to be an integer at all times.
3- Because this is legacy code I can't change the destination type to anything other than string
4- database/sql's conversion scheme is very restrictive on what can be mapped to what

A possible solution that comes to mind on the driver side is a generalized string map. Currently if this column is not mapped to an ora.OCINum it is mapped to a float, which would be fine if the string representation of the float isn' t like the following: 

`[sql: Scan error on column index 3: converting driver.Value type float64 ("2.738675e+06") to a int: invalid syntax]`

If there was a new ora number type on the driver side that is similar to ora.OCINum but returns custom strings based on the value of the number ( whether it has decimal points or not for example ) and instead of returning the type itself, it returned this string representation this would probably fix my issue. 

What do you think?

**Comment by omerkirk** [comment]
I will try the master as soon as I can, will let you know

**Comment by tgulacsi** [comment]
Be careful, thats "v4", not the "master" branch!

omerkirk <notifications@github.com> ezt írta (időpont: 2017. jan. 5., Cs,
16:11):

**Comment by omerkirk** [comment]
I see, is the "v4" branch only for go 1.8 or should I expect any breaking changes ?

**Comment by tgulacsi** [comment]
Some breaking changes, esp. The usage of ora.Cfg().

omerkirk <notifications@github.com> ezt írta (időpont: 2017. jan. 5., Cs
17:01):

**Comment by mynoheart** [comment]
```
var s string
	var i int64
	var f float64
	for tN, dest := range []interface{}{
		&f,
		&i,
		&s,
	} {
		if err := db.QueryRow("SELECT 0.018 FROM DUAL").Scan(dest); err != nil {
			fmt.Printf("%d. %T: %v", tN, dest, err)
		}
		fmt.Printf("%d. %T: %v", tN, dest, reflect.ValueOf(dest).Elem().Interface())
	}

outputs: 0. *float64: 0.0180000000000000021. *int64: sql: Scan error on column index 0: converting driver.Value type float64 ("0.018000000000000002") to a int64: invalid syntax1. *int64: 02. *string: 0.018000000000000002
```

I want return number as string (especially float's) in order not lost precision.

P.S.: 
```
drvCfg := ora.Cfg()
	drvCfg = drvCfg.SetNumberBigFloat(ora.S)
	drvCfg = drvCfg.SetNumberBigInt(ora.S)
	drvCfg = drvCfg.SetNumberFloat(ora.S)
	drvCfg = drvCfg.SetNumberInt(ora.S)
	ora.SetCfg(drvCfg)
```
I try and with this settings.

**Comment by mynoheart** [comment]
Sorry, i not set ```drvCfg.SetFloat(ora.S)```. With this settings work.

**Comment by omerkirk** [comment]
I finally upgraded our go version to go 1.9 and tried the trick you mentioned above. It works great for numbers now, however if a number column is null the string representation is "" which cannot be scanned to an int. I wonder how this works in goracle.v1.

**Comment by tgulacsi** [comment]
How would you scan a null into an int? What shall be the value in int?
Use sql.NullInt64 for this.

**Comment by omerkirk** [comment]
Yes you are right I just filed the issue since everything is working correctly in goracle.v1, but changed the query now and things are working ok in rana/ora.
